New codec obj and examples

jayteemo opened the pull request:
IRIS-6060

glarsennordic commented on include/net/nrf_cloud_codec.h:
```diff
@@ -20,6 +24,503 @@ extern "C" {
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
```
I wonder if __LAST would be useful here, too

jayteemo replied:
currently not useful, as the enum range is not checked.

glarsennordic commented on include/net/nrf_cloud_codec.h:
```diff
@@ -20,6 +24,503 @@ extern "C" {
+ * @retval -ENOTSUP Action not supported for the object's type.
+ * @retval 0 Success; item added.
+ */
+int nrf_cloud_obj_bulk_add(struct nrf_cloud_obj *const bulk, struct nrf_cloud_obj *const obj);
```
Since we distinguish the API by bulk vs. non-bulk message objects, I wonder if there should be a private boolean flag that tracks whether the object is a bulk message, so that the `nrf_cloud_obj_bulk` functions can sanity check the provided object?

glarsennordic replied:
Admittedly that might be over-optimizing. It would probably be sufficient to just use `nrf_cloud_obj_msg_check` for sanity checking instead of a flag

jayteemo replied:
the check for bulk is done inside `nrf_cloud_obj_bulk_add`:
```
if (!cJSON_IsArray(bulk->json)) {
	return -ENODEV;
}
```

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud.c:
```diff
@@ -348,11 +349,40 @@ int nrf_cloud_sensor_data_stream(const struct nrf_cloud_sensor_data *param)
+			}
+		}
+
+		/* Set object's encoded data for sending */
```
This comment is a touch misleading to me. I suggest `/* Collect object's encoded data for sending */`

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:
```diff
@@ -0,0 +1,805 @@
+
+LOG_MODULE_REGISTER(nrf_cloud_codec, CONFIG_NRF_CLOUD_LOG_LEVEL);
+
+int nrf_cloud_obj_input_decode(struct nrf_cloud_obj *const obj,
```
Maybe this function can accept a type hint, in case the receiver already knows to expect JSON or CBOR?

And then we can provide a wrapper to this function that has the logic to identify a message of unknown encoding

glarsennordic replied:
Actually, getting more specific, if you add the type hint, you could include sanity checks for each hinted type:

For JSON, the sanity checks could include:

1. Starts with `{` or `[`
2. Is NULL terminated
3. Parses successfully
4. nrf_cloud_obj_msg_check succeeds

For CBOR, the sanity checks could include:
1. Parses successfully
2. nrf_cloud_obj_msg_check succeeds

The wrapper could then sequentially call `nrf_cloud_obj_input_decode` with each possible encoding type until one succeeds

jayteemo replied:
what if the type hint was just the type of obj (`obj->type`) passed in?
if the type is undefined, it will try to decode the correct type.
if the type is defined, it will only try to decode that type.

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:
```diff
@@ -0,0 +1,805 @@
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		obj->json = NULL;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		return -ENOTSUP;
+	}
```
I wonder if we should instead raise an error if `obj->json` is not already NULL -- after all, that would suggest we are about to create a memory leak, wouldn't it?

Maybe we shouldn't support `nrf_cloud_obj_reset` at all? Sure, perhaps the user is doing something like taking ownership of the object and then discarding the `nrf_cloud_obj` struct, except they want to re-use it for another decode, but is that really a pattern we want to support?

Maybe there is a use case here that I'm missing, though

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:
```diff
@@ -0,0 +1,805 @@
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	if (obj->enc_src != NRF_CLOUD_ENC_SRC_PRE_ENCODED) {
+		return -EACCES;
+	}
+
+	if (obj->encoded_data.ptr) {
+		nrf_cloud_free((void *)obj->encoded_data.ptr);
+		obj->encoded_data.ptr = NULL;
+	}
+
+	obj->encoded_data.len = 0;
+	obj->enc_src = NRF_CLOUD_ENC_SRC_NONE;
+
+	return 0;
+}
+
+
+int nrf_cloud_obj_free(struct nrf_cloud_obj *const obj)
+{
+	if (!obj) {
+		return -EINVAL;
+	}
+
+	switch (obj->type) {
+	case NRF_CLOUD_OBJ_TYPE_JSON:
+	{
+		cJSON_Delete(obj->json);
+		obj->json = NULL;
+		return 0;
+	}
+	case NRF_CLOUD_OBJ_TYPE_CBOR:
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
```
I feel like all three of these could just be one function, `nrf_cloud_obj_free`, that checks whether there is encoded data that needs to be freed, as well as whether that data is pre-encoded or cloud-encoded

glarsennordic replied:
Actually this raises some questions about data lifetimes for pre-encoded objects in the first place. If the only way for pre-encoded objects to exist is for users to directly provide them, and if the only way such objects are freed is if the user requests that they be freed, perhaps it should just be up to the user to directly call free on the pre-encoded data they provide?

Also I'm a bit confused about `nrf_cloud_free`, it gets used for `pre-encoded` objects only ATM

It seems likely to me that I'm just suffering from a misunderstanding

jayteemo replied:
`nrf_cloud_obj_cloud_pre_encoded_free` was added just to be complete... maybe it doesn't make sense.  i used `nrf_cloud_free` so the proper free function would be used.

i feel that `nrf_cloud_obj_free` and `nrf_cloud_obj_cloud_encoded_free` should be separate functions as the memory is allocated by separate functions.  this gives the control to the user to free the specific item when they want.

glarsennordic replied:
> i feel that nrf_cloud_obj_free and nrf_cloud_obj_cloud_encoded_free should be separate functions as the memory is allocated by separate functions. this gives the control to the user to free the specific item when they want.

That makes sense, as long as we really do want to allow users to free just the encoded contents without freeing the whole nrf_cloud_obj

But do we really want that? I think there is an opportunity for cleanliness if we say "no, user, don't worry about the encoded data, just worry about freeing `nrf_cloud_obj` and we will keep track of the underlying encoded data for you"

Maybe I am wrong, though, and we really should offer that ability to the end user

glarsennordic replied:
Ah, thinking on it, your argument also makes more sense if we're talking about a stack-allocated `nrf_cloud_obj` as opposed to one that is heap-allocated.

Otherwise user might, in my suggestion, call `nrf_cloud_obj_cloud_encode` on a stack-allocated `nrf_cloud_obj` and it won't be nearly as obvious that doing so means they must then call `nrf_cloud_free` on the object. I catch your drift now.

I wonder, then, though: Does the user need access to `nrf_cloud_obj_cloud_encode`, or can this be something internal to the nrf_cloud lib? Privatizing it might make this lifetime issue simpler to design around

jayteemo replied:
currently `nrf_cloud_obj_cloud_encode` is only used internally by the nrf_cloud library.
so, yes it could be removed from the public interface.
however, i would lean towards leaving the capability publicly available in case there is some use-case not covered by the samples i've updated.  i think as long as it is documented properly, then it can stay public? 

i will remove `nrf_cloud_obj_cloud_pre_encoded_free.`.

glarsennordic commented on samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c:
```diff
@@ -497,20 +546,21 @@ int consume_device_message(void)
-		/* Re-enqueue the message for later retry (creating a new copy of it in heap). */
-		send_device_message(msg);
+		/* Re-enqueue the message for later retry */
+		enqueue_device_message(queued_msg, false);
```
If `enqueue_device_message` fails, here, the message object will never be freed

jayteemo replied:
ah good catch, i guess the return value of `enqueue_device_message` should be checked 🙈

glarsennordic commented on samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c:
```diff
@@ -541,57 +594,23 @@ int consume_device_message(void)
+		/* The message data now belongs to the queue.
+		 * Reset the provided object so it cannot be modified.
+		 */
+		nrf_cloud_obj_reset(msg_obj);
```
I guess this is the use case I was missing in my earlier comment.

But if the user is handing off ownership of the object to `connection.c`, enforcing that with `nrf_cloud_obj_reset` seems unnecessary to me, so `nrf_cloud_obj_reset` could still maybe be removed

jayteemo replied:
`nrf_cloud_obj_reset` is really just a convenience function for such a case.
the user could directly set the pointers to NULL if they want.

glarsennordic commented on include/net/nrf_cloud_codec.h:
```diff
@@ -20,37 +27,535 @@ extern "C" {
+	struct nrf_cloud_data encoded_data;
+};
+
+/** @brief Define an nRF Cloud JSON object.
+ *
+ * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_JSON.
+ *
+ * @param _name	Name of the object.
+ */
+#define NRF_CLOUD_OBJ_JSON_DEFINE(_name) \
+	struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_JSON, .json = NULL, \
+				       .enc_src = NRF_CLOUD_ENC_SRC_NONE, \
+				       .encoded_data = { .ptr = NULL, .len = 0 } }
```
I'm not 100% sure how I feel about this being a macro

On the one hand, it allows us to conveniently configure defaults, especially if any of them need to start out nonzero

On the other hand, I feel like user code would be more natural just straight up declaring `struct nrf_cloud_obj`

Overall I think I lean towards keeping the macros, but it still seems worthwhile to bring it up for discussion

jayteemo replied:
the macros were made to simplify struct initialization, so the user does not have to make sure they set each value correctly.
the user is free to declare and init the struct however they want.

jayteemo commented:
thanks for the review @glarsennordic 
i've addressed the comments, with https://github.com/jayteemo/fw-nrfconnect-nrf/pull/7#discussion_r1178492622 still being open... should `nrf_cloud_obj_cloud_encode` remain as is?
we can let @plskeggs add his thoughts on that item.

pete, i would also be interested in how (well) you see CBOR being implemented into this?  would any major API changes be needed or can the CBOR stuff just be dropped in to the NRF_CLOUD_OBJ_TYPE_CBOR cases.

plskeggs commented on applications/serial_lte_modem/src/gnss/slm_at_gnss.c:
```diff
@@ -574,28 +574,10 @@ static void send_location(struct nrf_modem_gnss_nmea_data_frame * const nmea_dat
-	cJSON_Delete(msg_obj);
-	if (json_msg) {
-		cJSON_free((void *)json_msg);
+	/* Encode the location data into a device message */
```
that's a lot cleaner

plskeggs commented on include/net/nrf_cloud_codec.h:
```diff
@@ -20,37 +26,530 @@ extern "C" {
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
```
Just so I understand, would this NRF_CLOUD_ENC_SRC_PRE_ENCODED be used with CBOR data that I encode using the CDDL, for example?

jayteemo replied:
it's mostly just a convenience type so that you can set some data in the obj and use the obj to send the data to the cloud... see usage of `NRF_CLOUD_OBJ_PRE_ENC_DEFINE` in  `slm_at_gnss.c`.

plskeggs commented on include/net/nrf_cloud_codec.h:
```diff
@@ -20,6 +23,498 @@ extern "C" {
+	union {
+		cJSON *json;
+		/* TODO:
+		 * struct zcbor_state_t cbor;
```
I looked at the code the zcbor code generator output, and learned that each specific use case needs one or more "backup states".  The generated code directly allocates on the stack an array of zcbor_state_t.  There are macros `ZCBOR_STATE_D` and `ZCBOR_STATE_E` for encoding and another for decoding which will allocate these for the user, using the formula:
`num_backups + 2`
where 'num_backups` is roughly how complex the overall message is.  See the not especially helpful docs: https://github.com/NordicSemiconductor/zcbor#the-zcbor-state-object.

I wonder if it makes sense to declare here in this union just a pointer to one or more zcbor_state_t.  This could point to a dynamically allocated array that would be freed when `nrf_cloud_obj_free()` is called.

plskeggs commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:
```diff
@@ -0,0 +1,814 @@
+	return -ENOTSUP;
+}
+
+int nrf_cloud_obj_msg_init(struct nrf_cloud_obj *const obj, const char *const app_id,
```
If we have the zcbor section allocate an array of N zcbor_state_t, then this function will need a parameter to set N which would be unused for JSON.

jayteemo replied:
how about handling the N value as follows:
```
struct nrf_cloud_cbor {
	uint_fast32_t states;
	/* TODO:
	struct zcbor_state_t *zcb;
	*/
};

/** @brief Object used for building nRF Cloud messages. */
struct nrf_cloud_obj {

	enum nrf_cloud_obj_type type;
	union {
		cJSON *json;
		struct nrf_cloud_cbor cbor;
	};

	/** Source of encoded data */
	enum nrf_cloud_enc_src enc_src;
	/** Encoded data */
	struct nrf_cloud_data encoded_data;
};

/** @brief Define an nRF Cloud CBOR object.
 *
 * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_CBOR.
 *
 * @param _name	Name of the object.
 */
#define NRF_CLOUD_OBJ_CBOR_DEFINE(_name, _states) \
struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_CBOR, \
				.enc_src = NRF_CLOUD_ENC_SRC_NONE, \
				.encoded_data = { .ptr = NULL, .len = 0 } \
				.cbor = { .states = _states } }

```

plskeggs replied:
In this scenario, who allocates and frees the `states` count of `struct zcbor_state_t`?

jayteemo replied:
it would be allocated in `nrf_cloud_obj_init ` (or `nrf_cloud_obj_msg_init `) and freed in `nrf_cloud_obj_free`

plskeggs replied:
OK, sounds good to me.
